### PR TITLE
PoC: Add support for DML with properties in engine + Add support for branching in Iceberg

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

The 1st commit allows properties after the target table name in DML:
```g4
    | INSERT INTO qualifiedName (WITH properties)?
       columnAliases? rootQuery                                        #insertInto
    | DELETE FROM qualifiedName (WITH properties)?
         (WHERE booleanExpression)?                                    #delete
    | UPDATE qualifiedName (WITH properties)?
        SET updateAssignment (',' updateAssignment)*
        (WHERE where=booleanExpression)?                               #update
    | MERGE INTO qualifiedName (AS? identifier)? (WITH properties)?
        USING relation ON expression mergeCase+                        #merge
```

Avoided session property approach because users may forget the current branch and modify unexpected branch. 

The 2nd commit introduces `target_branch` property to get a branch name in DML and adds the following procedures in Iceberg connector:
* ALTER TABLE table EXECUTE create_branch(name)
  * creating a new branch
* ALTER TABLE table EXECUTE drop_branch(name)
  * removing a branch by name
* ALTER TABLE table EXECUTE fast_forward(from, to) 
  * performs a fast-forward of `from` up to the `to` snapshot if `from` is an ancestor of `to`.

TestIcebergBranching uses those new syntax and procedures.

Relates to #12844

References
* https://iceberg.apache.org/docs/latest/branching/
* https://docs.dremio.com/current/reference/sql/commands/nessie/

ToDo
- [ ] Deprecated old methods in SPI
- [ ] Investigate the expected behavior when modifying branch with schema evolution in Iceberg
- [ ] Fix handling of table statistics 

## Release notes

```markdown
# SPI
* TBD. ({issue}`issuenumber`)

# Iceberg
* Add support for `create_branch`, `drop_branch` and `fast_forward` table procedures. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            assertBranch(table.getName(), "main");
+
+            createBranch(table.getName(), "test-branch");
+            assertUpdate("INSERT INTO " + table.getName() + " WITH (target_branch = 'test-branch') VALUES 1, 2, 3", 3);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName()))
+                    .isEqualTo(0L);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName() + " FOR VERSION AS OF 'test-branch'"))
+                    .isEqualTo(3L);
+
+            fastForward(table.getName(), "main", "test-branch");
```
Comment (pajaks) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java:103]: Could you add test for case with changes on main after creating test-branch.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName()))
+                    .isEqualTo(3L);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName() + " FOR VERSION AS OF 'test-branch'"))
+                    .isEqualTo(3L);
+
+            // Iceberg creates a new branch if "from" branch does not exist
+            fastForward(table.getName(), "new-branch", "test-branch");
```
Comment (pajaks) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java:110]: What should be expected outcome if we tried to fast forwad to existing not related branch (not main)

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName() + " FOR VERSION AS OF 'test-branch'"))
+                    .isEqualTo(3L);
+
+            fastForward(table.getName(), "main", "test-branch");
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName()))
+                    .isEqualTo(3L);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName() + " FOR VERSION AS OF 'test-branch'"))
+                    .isEqualTo(3L);
+
+            // Iceberg creates a new branch if "from" branch does not exist
```
Comment (pajaks) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java:109]: Should we behave in the same way, or should we throw here?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        }
+    }
+
+    @Test
+    void testMerge()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_update_branch", "(x int)")) {
+            createBranch(table.getName(), "test-branch");
+
+            assertUpdate("MERGE INTO " + table.getName() + " WITH (target_branch = 'test-branch') USING (VALUES 42) t(dummy) ON false " +
```
Comment (pajaks) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java:197]: Can you add also merge with update case?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                    .isEqualTo(3L);
+
+            assertQueryFails(
+                    "ALTER TABLE " + table.getName() + " EXECUTE fast_forward('main', 'non-existng-branch')",
+                    "Branch 'main' is not an ancestor of 'non-existng-branch'");
+        }
+    }
+
+    @Test
+    void testInsert()
```
Comment (pajaks) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java:121]: Can you add tests for insert (and other updates) into branch after DDL change on main?
It should be possible probably, currently it's failing like:
`Insert query has mismatched column types: Table: [integer, integer, integer], Query: [integer, integer]`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 6]
```diff
@@ -1825,6 +1901,64 @@ private void removeOrphanFiles(Table table, ConnectorSession session, SchemaTabl
...
+    public void executeFastForward(ConnectorSession session, IcebergTableExecuteHandle executeHandle)
+    {
+        IcebergFastForwardHandle fastForwardHandle = (IcebergFastForwardHandle) executeHandle.procedureHandle();
+        String from = fastForwardHandle.from();
+        String to = fastForwardHandle.to();
+
+        Table table = catalog.loadTable(session, executeHandle.schemaTableName());
+        try {
+            table.manageSnapshots()
+                    .fastForwardBranch(from, to)
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1951]: Not yet, but I suppose we should provide both for avoiding breaking changes in the future.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+            // insert into another branch with a partial column
+            assertUpdate("INSERT INTO " + table.getName() + " WITH (target_branch = 'test-branch') (x) VALUES 50", 1);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName()))
+                    .isEqualTo(1L);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName() + " FOR VERSION AS OF 'test-branch'"))
+                    .isEqualTo(3L);
+
+            // change table definition on main branch
+            assertUpdate("ALTER TABLE " + table.getName() + " ADD COLUMN z int");
+            assertUpdate("INSERT INTO " + table.getName() + " WITH (target_branch = 'test-branch') SELECT x, y, NULL FROM " + table.getName(), 1);
```
Comment (pajaks) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java:162]: Are you sure that branch should be forced to be used with schema from current main? What if user drop column from main branch and we still want to add data for this column on branch? What if we want select with predicate on dropped column?

**Comment by ebyhr** [comment]
Opened a new PR with `@` syntax. https://github.com/trinodb/trino/pull/24556

**Pull request closed without merging**